Fix notification reset events potentially arriving out of order if a block operation times out

peppy opened the pull request:
Inline comments should explain the issue.

I'm... not sure this is worth writing test coverage for (going to be very high bar to repro the scenario, I think).

- [x] Depends on https://github.com/ppy/osu-framework/pull/5002

peppy committed:
Fix notification reset events potentially arriving out of order if a block operation times out

peppy committed:
Remove synchronization context hacks in realm tests

peppy commented:
Added dependency on framework change in order to avoid smashing my head on keyboard attempting to make tests work with hacks.

peppy committed:
Use blocking calls to `SynchronizationContext` to guarantee order of execution

peppy commented:
I've updated this in line with the changes in https://github.com/ppy/osu-framework/pull/5002. Together they should give sane behaviour.

bdach commented on osu.Game/Database/RealmAccess.cs:
```diff
@@ -605,16 +586,33 @@ public IDisposable BlockAllOperations()
+                // calls above.
+                syncContext?.Send(_ =>
+                {
+                    foreach (var action in notificationsResetMap.Values)
+                        action();
```
This is the only usage of `notificationResetMap` which doesn't take `realmLock`. Same as last time, it probably doesn't strictly matter as all of the surrounding code is pretty much only allowed to run on the update thread, but given that this sort of thing was pointed out in review last time [no less](https://github.com/ppy/osu/pull/16547#discussion_r790278090) [than twice](https://github.com/ppy/osu/pull/16547#discussion_r790505471) then maybe better to just lock for the sake of it?

bdach commented on osu.Game/Database/RealmAccess.cs:
```diff
@@ -649,8 +647,14 @@ void restoreOperation()
+                // In the case we are on the update thread, let's also require this to run synchronously.
+                // This requirement is mostly due to test coverage, but shouldn't cause any harm.
+                if (ThreadSafety.IsUpdateThread)
+                    syncContext?.Send(_ => ensureUpdateRealm(), null);
```
So I guess these `Send()` calls and their inner "flush the scheduler" calls are supposed to enforce that the order of execution of these realm callbacks is consistent with the order in which they were posted/sent - but in that case, should the following also use `Send()` in the update thread branch?

https://github.com/ppy/osu/blob/4a9f4eecba45e66428de23094f1d1387199bb450/osu.Game/Database/RealmAccess.cs#L298-L301

Because that looks like the mirror image of this situation.

peppy replied:
It shouldn't matter here as well, both will provide non-incorrect behaviour. which do you think reads better?

to rephrase, both could be direct calls of that is preferred.

bdach replied:
I actually think I prefer using `.Send()` in both cases, having the guarantee of in-enqueue-order execution seems valuable. But that's not a super strong opinion, I'm mostly looking for consistency.

bdach reviewed (commented):
Generally the idea of ensuring that the subscriptions are dead before invoking the reset events looks correct, just a few tiny details that I'm not sure about yet.

peppy committed:
Merge branch 'master' into fix-out-of-order-events-on-block-fail

peppy committed:
Update framework

peppy committed:
`Send` unsubscribe actions to synchronization context for consistency and safety

smoogipoo commented:
Block button now permanently deadlocks osu!.

peppy committed:
Move final result set firing to before the update realm is disposed

Without this, if any registered callback attempts to access
`RealmAccess.Realm` when handling the empty set callback, it will
deadlock the game.

peppy committed:
Revert "Move final result set firing to before the update realm is disposed"

This reverts commit 24bcba64183a7bf404eca503a3c31a24762bfdac.

peppy committed:
Add flag to guard against deadlocks during blocking operations

peppy committed:
Move final empty result set sending to post-compact

frenzibyte commented:
Looks to have actual failures on Windows CI.

peppy commented:
Those are fine, is a known but unrelated issue. We need retry logic on the migration operation. There's no way to know when the file handle is closed from our end.

smoogipoo commented:
Going to guess the windows failures are actually caused by this branch, because they don't seem to happen anywhere else.

peppy commented:
I'm doing some investigation locally, will see if I can reproduce.

peppy committed:
Fix completely invalid method of testing realm migration

peppy commented:
The reason for the failure is that with this PR, we are now more correctly and more synchronously restoring the update realm instance. The way the test was written, this would just not work. It only worked by chance until now because we were doing things wrong.

Latest commit fixes over 20+ runs on windows.

smoogipoo committed:
Merge branch 'master' into fix-out-of-order-events-on-block-fail
